src/main/java/com/google/sps/servlets/startGameServlet.java

```diff
@@ -0,0 +1,85 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+ 
+package com.google.sps.servlets;
+ 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.sps.utils.GameUtils;
+import com.google.sps.utils.UserUtils;
+import java.io.*;
+import javax.servlet.*;
+import javax.servlet.http.*;
+import java.io.PrintWriter;
+ 
+/** Create game entity and connect it to the user, then update the user entity to include the game id
+    and store both in datastore when the user creates a new game.
+ */
+@WebServlet("/start-game")
+public class startGameServlet extends HttpServlet {
+ 
+  DatastoreService datastore;
+ 
+  public startGameServlet(){
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+ 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+ 
+    String gameName = request.getParameter("game-name");
+    
+    // query user 
+    Cookie cookies[] = request.getCookies();
+    Entity userEntity = UserUtils.getUserFromCookie(cookies, datastore);
+ 
+    if(userEntity == null){
+        // no user entity, send back to login
+        response.sendRedirect("/index.html");
+        return;
+    }
+ 
+    // prevent users from joining more than one game
+    if((String) userEntity.getProperty("gameId") != null){
+      PrintWriter out = response.getWriter();
+      out.println("<p>You are already in a game, see the gameboard here:</p>");
+      out.println("<a href = \"gameBoard.html\"><h3>Your game</h3></a>");
+      return;
+    }
+      
+    // create game entity
+    Entity gameEntity = GameUtils.createGameEntity(gameName, datastore);
+ 
+    // add user to game entity + vice versa
+    boolean userAdded = false;
+    boolean gameAdded = false;
+    if(gameEntity != null){
+        userAdded = GameUtils.addUserToGame((String) userEntity.getProperty("userId"), gameEntity, datastore);
+        gameAdded = UserUtils.addGameToUser(userEntity, datastore, (String) gameEntity.getProperty("gameId"));
+    }
+    
+    if(userAdded && gameAdded){
+        response.sendRedirect("/gameBoard.html");
+    }else{
+        // something's wrong, send back to login
+        response.sendRedirect("/index.html");
+    }
```

Comment:
This is a pretty vague else clause. We have no idea what had to happen for us to get here. Is userAdded False? Is gameAdded False? What does it mean if either of those are false?

This would be a good place to do some logging. I would suggest a little rework of the flow from getting the gameEntity to the end of file. First, just return early if gameEntity is null rather than checking to see if it's not null. Then, check the status of addUserToGame(). If it's False, log why then redirect to index.html. Then do the same for addGameToUser. At this point, we've checked all there is to check so we can safely just redirect to gameBoard.html without needed to do any more checks.

Reply:
This is all broken up in the helper function